src/api/js/src/high-level/types.ts

```diff
@@ -0,0 +1,415 @@
+
+  add(other: ArithRef<Name> | number | bigint): ArithRef<Name>;
+  mul(other: ArithRef<Name> | number | bigint): ArithRef<Name>;
+  sub(other: ArithRef<Name> | number | bigint): ArithRef<Name>;
+  pow(exponent: ArithRef<Name> | number | bigint): ArithRef<Name>;
+  div(other: ArithRef<Name> | number | bigint): ArithRef<Name>;
+  mod(other: ArithRef<Name> | number | bigint): ArithRef<Name>;
+  neg(): ArithRef<Name>;
+  le(other: ArithRef<Name> | number | bigint): BoolRef<Name>;
+  lt(other: ArithRef<Name> | number | bigint): BoolRef<Name>;
+  gt(other: ArithRef<Name> | number | bigint): BoolRef<Name>;
+  ge(other: ArithRef<Name> | number | bigint): BoolRef<Name>;
+}
+
+export interface Probe<Name extends string = string> {
+  readonly __typename: 'Probe';
+
+  readonly ctx: Context<Name>;
+  readonly probe: Z3_probe;
+}
+
+export interface TacticCtor<Name extends string> {
+  new (name: string): Tactic<Name>;
+}
+export interface Tactic<Name extends string = string> {
+  readonly __typename: 'Tactic';
+
+  readonly ctx: Context<Name>;
+  readonly ptr: Z3_tactic;
+}
+
+export interface AstVectorCtor<Name extends string> {
+  new <Item extends AstRef<Name> = AnyAst<Name>>(): AstVector<Item, Name>;
+}
+export interface AstVector<Item extends AstRef = AnyAst, Name extends string = string> extends Iterable<Item> {
+  readonly __typename: 'AstVector';
+
+  readonly ctx: Context<Name>;
+  readonly ptr: Z3_ast_vector;
+  get length(): number;
+
+  entries(): IterableIterator<[number, Item]>;
+  keys(): IterableIterator<number>;
+  values(): IterableIterator<Item>;
+  get(i: number): Item;
+  get(from: number, to: number): Item[];
+  set(i: number, v: Item): void;
+  push(v: Item): void;
+  resize(size: number): void;
+  has(v: Item): boolean;
+  sexpr(): string;
+}
+
+export interface AstMapCtor<Name extends string> {
+  new <Key extends AstRef = AnyAst, Value extends AstRef = AnyAst>(): AstMap<Key, Value, Name>;
+}
+export interface AstMap<Key extends AstRef = AnyAst, Value extends AstRef = AnyAst, Name extends string = string> {
+  readonly __typename: 'AstMap';
+
+  readonly ctx: Context<Name>;
+  readonly ptr: Z3_ast_map;
+}
+
+export type CoercibleToExpr<Name extends string = string> = number | bigint | boolean | ExprRef<Name>;
+
+type QuantifierOptions = {
+  qid: string;
+  skid: string;
+  patterns: PatternRef[];
+  noPatterns: ExprRef[];
+};
+
+export type Z3HighLevel = {
+  // Global functions
+  enableTrace(tag: string): void;
+  disableTrace(tag: string): void;
+  getVersion(): {
+    major: number;
+    minor: number;
+    build_number: number;
+    revision_number: number;
+  };
+  getVersionString(): string;
+  getFullVersion(): string;
+  openLog(filename: string): boolean;
+  appendLog(s: string): void;
+  setParam(key: string, value: any): void;
+  setParam(key: Record<string, any>): void;
+  resetParams(): void;
+  getParam(name: string): string | null;
+
+  // Operations that require context
+  createContext(): Z3WithContext<string>;
```

Comment:
`string` is the supertype of all strings, so making this default to `string` will prevent any useful context-based typechecking. E.g., it means that `api.createContext().And(api.createContext('other').Bool('x'))` will typecheck. The type system can and should prevent that from working.

If the name isn't required then no one is ever going to know it exists, whereas if the name is required it's a lot more obvious and not any harder to use. Do you think it's important to allow unnamed contexts?

Reply:
You're right, skipping the whole type system is not a great idea, I'll backtrack on that.